timelock-corruption-detection/src/main/java/com/palantir/timelock/history/PaxosLogHistoryProvider.java

```diff
@@ -61,16 +77,189 @@ private Long getOrInsertVerificationState(NamespaceAndUseCase namespaceAndUseCas
                 namespaceAndUseCase.namespace(), namespaceAndUseCase.useCase());
     }
 
+    /**
+     * Gets history from all the nodes for all unique (namespace, useCase) tuples since last verified sequence numbers
+     * i.e. the highest sequence number that was verified since the last time the bounds were reset.
+     *
+     * @throws Exception if fails to fetch history from all remote servers
+     */
     public List<CompletePaxosHistoryForNamespaceAndUseCase> getHistory() {
-        Map<NamespaceAndUseCase, Long> lastVerifiedSequences = KeyedStream
+
+        /**
+         * map of all unique (namespace, useCase) tuples to their respective last verified sequence numbers.
+         */
+        Map<NamespaceAndUseCase, Long> lastVerifiedSequences = getNamespaceAndUseCaseToLastVerifiedSeqMap();
+        PaxosHistoryOnSingleNode localPaxosHistory = localHistoryLoader.getLocalPaxosHistory(lastVerifiedSequences);
+
+        /**
+         * The history queries are built from the lastVerifiedSequences map above,
+         * required for conjure endpoints to load remote history.
+         */
+        List<HistoryQuery> historyQueries = getHistoryQueryListForRemoteServers(lastVerifiedSequences);
+
+        /**
+         * List of logs from all remotes.
+         */
+        List<PaxosHistoryOnRemote> rawHistoryFromAllRemotes = getHistoriesFromRemoteServers(historyQueries);
+
+        /**
+         * List of consolidated histories from all remotes. Each history is a map of namespaceAndUseCase
+         * to {@link ConsolidatedLearnerAndAcceptorRecord}.
+         */
+        List<NamespaceAndUseCaseWiseConsolidatedLearnerAndAcceptorRecords> historyFromAllRemotes
+                = buildHistoryFromRemoteResponses(rawHistoryFromAllRemotes);
+
+        /**
+         * Consolidate and build complete history for each (namespace, useCase) pair
+         * from histories loaded from local and remote servers.
+         */
+        return consolidateAndGetHistoriesAcrossAllNodes(
+                lastVerifiedSequences,
+                localPaxosHistory,
+                historyFromAllRemotes);
+    }
```

Comment:
This is quite a heavy class and it feels like the last two parts on translating `PaxosHistoryOnRemote` to `CompletePaxosHistoryForNamespaceAndUseCase` seems a bit of a separate concern from getting the data which the first three parts here do. We should split this up, though I'm fine with getting this in first as long as we track doing this when we return to this project later on.